### PR TITLE
Add support for auto-generating multiple RCs for the current version if a previous RC exists

### PR DESCRIPTION
**Pull request by fullofcaffeine**
## Description

Corresponding issue: https://github.com/WordPress/gutenberg/issues/25970

**Latest update and test results here:** https://github.com/WordPress/gutenberg/pull/25971#issuecomment-708797141.

Add support for auto-generating multiple RCs or the current version if a previous RC exist.

If a previous RC exists, it will generate a new RC, in sequence. If not, it will fallback to the regular logic of generating an RC1 for the `nextMajor` version, as well as creating the new `releaseBranch`.

## How has this been tested?

I've run it locally and changed the config to point to my fork of GB. 

To test it out, change the values of https://github.com/WordPress/gutenberg/blob/master/bin/plugin/config.js to point to a test fork of Gutenberg. [Example](https://gist.github.com/fullofcaffeine/2ed4ac9d8fa219d5072defdf9aeac31b).

1. First, run `./bin/plugin/cli.js rc` from Gitenberg's root dir. This assumes the last release is a stable non-RC one. The behaviour here should be the same as before the changes here.

1. Now, you can take note of where the script cloned the version of GB downloaded when you ran the script in 1), and then manually set this var to point to its path: https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/commands/release.js#L719 (i.e: `const gitWorkingDirectory = '/tmp/59506893-74ff-4012-9486-36a0c7ca2869';`)

1. It will then release RC1 for the next major version;

1. Run the script again and it should pick up the current version and will automatically increment it to RC2. 

1. Run `./bin/plugin/cli.js stable` to release a stable version.

1. Run it again to release a point release.

## Types of changes

New feature (non-breaking change which adds functionality) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: bin/plugin/commands/release.js [hunk 1]
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
 ) {
 	let version, releaseBranch, versionLabel;
 	await runStep( 'Creating the release branch', abortMessage, async () => {
+		let parsedMajorVersion;
+
+		const getComponentsOfRC = ( ver ) => ver.match( /(.+-rc\.)([0-9]+)/ );
```
Comment (ockham) [inline bin/plugin/commands/release.js:76]: I think we can use `semver.parse()` instead, and inspect the resulting `prerelease` attribute. Something like

```js
const packageVersion = semver.parse( packageJson.version );
const { prerelease } = packageVersion;
```

File: bin/plugin/commands/release.js [hunk 2]
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
...
 		const packageJsonPath = gitWorkingDirectoryPath + '/package.json';
 		const packageJson = readJSONFile( packageJsonPath );
-		const nextMajor = getNextMajorVersion( packageJson.version );
-		const parsedMajorVersion = semver.parse( nextMajor );
+		const rcComponents = getComponentsOfRC( packageJson.version );
+		const isRC = !! rcComponents;
```
Comment (ockham) [inline bin/plugin/commands/release.js:80]: While `isRC` is accurate naming, it's prone to confusion with the (user-supplied) `isRC` bool flag in `releasePlugin()`. Let's maybe rename this one. `packageVersionIsRC` maybe?

Comment (ockham) [reply bin/plugin/commands/release.js:80]: or `isPackageVersionRC`? That's more predicate-style.

File: bin/plugin/commands/release.js (hunk 1 above)
Comment (ockham) [reply bin/plugin/commands/release.js:76]: Example:

```
const version = semver.parse( '9.1.2-rc.1' );
console.log( version.prerelease );
```

gives `[ 'rc', 1 ]` :tada:

File: bin/plugin/commands/release.js [hunk 3]
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
...
+
+		// if the last tag is an RC then we can assume that we should generate a new rc with the same version number
```
Comment (ockham) [inline bin/plugin/commands/release.js:82]: :+1: :100:

File: bin/plugin/commands/release.js [hunk 4]
```diff
@@ -71,34 +71,64 @@ async function runReleaseBranchCreationStep(
...
+		if ( isRC ) {
+			const [ , current, rc ] = rcComponents;
+			const nextRC = Number( rc ) + 1;
+
+			version = `${ current }${ nextRC }`;
+			versionLabel = `${ current.split( '-' )[ 0 ] } RC${ nextRC }`;
+			parsedMajorVersion = semver.parse( packageJson.version );
```
Comment (ockham) [inline bin/plugin/commands/release.js:89]: We can use `semver.inc()` for the version arithmetics here:

```js
version = semver.inc( packageVersion, 'prerelease' );
```

**Review by ockham — COMMENTED** [review]
Left a few notes. Looking pretty good overall!

**Comment by ockham** [comment]
We should also update this line: https://github.com/WordPress/gutenberg/blob/d4a74cca63617405b2bf2680ef8011dfb87dd367/bin/plugin/cli.js#L34

File: bin/plugin/commands/release.js (hunk 1 above)
Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:76]: Ah, good one! I'll change 👍🏻

File: bin/plugin/commands/release.js (hunk 2 above)
Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:80]: Ohh, I missed that bool flag. I think `isPackageVersionRC` is a good name 👍🏻

File: bin/plugin/commands/release.js (hunk 4 above)
Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:89]: TIL 👍🏻

File: bin/plugin/cli.js [hunk 5]
```diff
@@ -31,7 +31,7 @@ program
 	.command( 'release-plugin-rc' )
 	.alias( 'rc' )
 	.description(
-		'Release an RC version of the plugin (supports only rc.1 for now)'
+		'Release an RC version of the plugin (will increment the RC number if one already exists for the last version)'
```
Comment (ockham) [inline bin/plugin/cli.js:34]: We can probably drop the note in parens altogether, I think it's kinda implied, and we're doing the same for stable versions where we also don't mention the increment.

```suggestion
		'Release an RC version of the plugin'
```

File: bin/plugin/commands/release.js [hunk 6]
```diff
@@ -69,40 +69,59 @@ async function runReleaseBranchCreationStep(
...
+		if ( isPackageVersionRC ) {
+			parsedVersion = semver.parse(
+				semver.inc( packageVersion.version, 'prerelease', 'rc' )
+			);
+		} else {
+			const nextMajor = getNextMajorVersion( packageJson.version );
+			parsedVersion = semver.parse( nextMajor + '-rc.1' );
+		}
+
+		versionLabel = parsedVersion.raw.replace( /\-rc\.([0-9]+)/, ' RC$1' );
```
Comment (ockham) [inline bin/plugin/commands/release.js:89]: Let's maybe use a (non-parsed) `version` as the variable, since we need that more often than the parsed one:

```suggestion
		if ( isPackageVersionRC ) {
			version = semver.inc( packageVersion.version, 'prerelease', 'rc' );
		} else {
			const nextMajor = getNextMajorVersion( packageJson.version );
			version = nextMajor + '-rc.1';
		}

		versionLabel = version.replace( /\-rc\.([0-9]+)/, ' RC$1' );
```

**Review by ockham — APPROVED** [review]
LGTM, and tests well (as far as I could test).

Let's wait for tests to go green, and ideally someone who's more familiar with these tools to approve :+1: 

cc/ @youknowriad @gziolo @mcsf

File: bin/plugin/commands/release.js [hunk 7]
```diff
@@ -390,9 +387,11 @@ async function runPushGitChangesStep(
 				true,
 				abortMessage
 			);
+			// We need to force-push because we are effectivelly rebasing from master
```
Comment (fullofcaffeine) [inline bin/plugin/commands/release.js:390]: Since the release is cherry-picked into master, the SHA changes, thus, pushing to the release branch is not possible anymore without force-pushing. 

This seems to work well, but I don't know if there might be unforeseen side-effects to force-pushing to the release branch. I'd love some feedback here.

File: bin/plugin/lib/git.js [hunk 8]
```diff
@@ -82,10 +82,20 @@ async function createLocalTag( gitWorkingDirectoryPath, tagName ) {
...
-async function pushBranchToOrigin( gitWorkingDirectoryPath, branchName ) {
+async function pushBranchToOrigin(
+	gitWorkingDirectoryPath,
+	branchName,
+	options = []
+) {
 	const simpleGit = SimpleGit( gitWorkingDirectoryPath );
-	await simpleGit.push( 'origin', branchName );
+
+	// TODO The type definitions for simple-git are wrong and will complain here.
```
Comment (fullofcaffeine) [inline bin/plugin/lib/git.js:94]: As the comment states, there seems to be a mismatch between the TS typing declarations and the js version of simple-git. I tried passing the `options` array as the last arg as the type docs say, but it didn't work. 

After inspecting the actual code, I realized thet last param is called "then" and is unused. By checking the implementation, I saw that the right way to pass these options was to pass them as the first argument.

I'll look into it again tomorrow, but if anyone else more acquainted with the codebase has any insights, please let me know :)

Comment (fullofcaffeine) [reply bin/plugin/lib/git.js]: I managed to fix this by passing the options as an object instead of an array (as the last argument). The interface here is not the most intuitive ¯\_(ツ)_/¯

**Comment by fullofcaffeine** [comment]
This seems to work well. I just tested twice in order to get RC1 and the subsequent RC2:

1. Output for RC1: https://gist.github.com/fullofcaffeine/848e637f16c0b856766185620d9574ee
2. Output for RC2: https://gist.github.com/fullofcaffeine/d0a2880ac430bc60d40d67547faee24c

File: bin/plugin/commands/release.js (hunk 7 above)
Comment (youknowriad) [reply bin/plugin/commands/release.js:390]: I don't like this. we shouldn't be force pushing to master IMO, we could potentially lose merges that happened in the meantime.

**Comment by youknowriad** [comment]
This is cool, aside the force push which I think is very risky,  this is looking good.

The pre-release is normal, we mark all "RCs" as "pre-releases" explicitly in the script.

File: bin/plugin/commands/release.js [hunk 9]
```diff
@@ -73,32 +73,48 @@ async function runReleaseBranchCreationStep(
...
+				true,
+				abortMessage
+			);
 
-		// Creating the release branch
+			log(
+				'>> The local release branch ' +
+					formats.success( releaseBranch ) +
+					' has been successfully created.'
+			);
```
Comment (mcsf) [inline bin/plugin/commands/release.js:106]: Shouldn't this remain after the call to `git.createLocalBranch`?

File: bin/plugin/commands/release.js (hunk 7 above)
Comment (mcsf) [reply bin/plugin/commands/release.js:390]: At the very least, this should be [`--force-with-lease`](https://git-scm.com/docs/git-push).

(Btw, everyone should use that option in their daily Git interactions — I recommend adding aliases to Git like ` fpush = push --force-with-lease` or `fp` or `pf` or whatever.)

Comment (ockham) [reply bin/plugin/commands/release.js:390]: Our ideal scenario here would be to have all RC commits (alongside potential other, cherry-picked, ones) on the release branch, no? So we need to ensure that we start off from the point where we "left off" with the previous RC. I'm wondering what's currently happening that gets in the way of that?

Could it be that the `git.checkoutLocalBranch` call in `runReleaseBranchCreationStep` should only be called if the branch didn't exist before, and that we need to run `checkoutRemoteBranch` otherwise?

Comment (ockham) [reply bin/plugin/commands/release.js:390]: > Since the release commit is cherry-picked into master, the SHA changes, thus, pushing to the release branch is not possible anymore without force-pushing in the case of subsequent RCs.

Are you sure that's the reason? :thinking: Wouldn't we run into the same problem when releasing a stable version after an RC?

Comment (ockham) [reply bin/plugin/commands/release.js:390]: I'm starting to think it would make sense to change the criterion for the ternary that runs either `runReleaseBranchCreationStep` or `runReleaseBranchCheckoutStep` from `isRC` to `isPackageVersionRC`, since it captures better what we're trying to do here:

- If the `package.json` version is an RC, then we can assume that a release branch has been created.
- If it doesn't, we're at the start of a new release cycle and need to create that branch.

This will mean moving some of the version computing logic around, but I think it'll fall in place quite naturally.

Comment (youknowriad) [reply bin/plugin/commands/release.js:390]: @ockham that makes sense since second/third RCs will just just use the existing branch with cherrypicked commits

Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:390]: > I don't like this. we shouldn't be force pushing to master IMO, we could potentially lose merges that happened in the meantime.

Note that the force push here is not to master, but to the `releaseBranch`.

File: bin/plugin/commands/release.js (hunk 9 above)
Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:106]: Good point. The call to `createLocalBranch` has been moved down becuase I found out that was also needed for the RC scenario, as it essentially checkout th branch. It doesn't make sense indeed to show a confirmation message before the actual thing happened :)

**Comment by fullofcaffeine** [comment]
Thanks @ockham for lending your hand today and to help me to understand the underlying issue (that led me to try the force-push approach*) 🙇🏻 

I had to re-arrange the logic slithgly again as it was failing just when trying to update from stable to stable. I tried to simplify it a bit, too. 

I did another round of tests (using my own fork of GB, see config [here](https://gist.github.com/fullofcaffeine/2ed4ac9d8fa219d5072defdf9aeac31b)) and it now seems to work well (no more force-push needed). Here are the scenarios and their respective results:

|  command               | package version | new version | observations |
| ------------- | ------------- | --- | --- |
| rc | `9.1.1`  | `9.2.0-rc.1` | Uses [this](https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/lib/version.js#L13) function to generate the next version |
| rc  | `9.2.0-rc.1`  | `9.2.0-rc.2` | increments the RC ad infinitum until a stable version is released |
| stable | `9.2.0-rc.2` | `9.2.0` | increments the patch version, effectively removing the RC portion |
| stable | `9.2.0` | `9.2.1` | increments the patch version ad infinitum until we release another RC* |

_Note that I didn't fully finish the whole process when running the `stable` command, I replied 'N' when it got to the SVN part, but I don't think it's needed to consider these tests as succesful, as all the relevant steps have been run prior to it._

*This effectively makes the `rc` command mandatory for new [minor/major](https://github.com/fullofcaffeine/gutenberg/blob/add/support-multiple-rcs-release-tool/bin/plugin/lib/version.js#L13) versions.

_*Basically, we were trying to create a branch and push in place of checking out the existing branch for the scenarios that created a release in the same major/minor version. We had to adjust the logic to be aware of that, too._

File: bin/plugin/commands/release.js [hunk 10]
```diff
@@ -189,7 +146,6 @@ async function runReleaseBranchCheckoutStep(
 
 	return {
 		version,
```
Comment (ockham) [inline bin/plugin/commands/release.js:148]: Let's drop `version` from the returned object -- we're expecting it as a function arg now and aren't modifying it, so there's no point really to return it.

File: bin/plugin/commands/release.js [hunk 11]
```diff
@@ -60,123 +60,80 @@ async function runSvnRepositoryCheckoutStep( abortMessage ) {
...
-		}
+		log(
+			'>> The local release branch ' +
+				formats.success( releaseBranch ) +
+				' has been successfully created.'
+		);
 	} );
 
 	return {
 		version,
```
Comment (ockham) [inline bin/plugin/commands/release.js:99]: Same as https://github.com/WordPress/gutenberg/pull/25971/commits/069f5f8cc1c0e6b3d981ad74744f8b9ee81b1fd1#r505636588

File: bin/plugin/commands/release.js [hunk 12]
```diff
@@ -720,21 +678,39 @@ async function releasePlugin( isRC = true ) {
...
+		// If packageVersion is stable, then checkout the branch and inc patch.
+		// If packageVersion is RC, then checkout the branch and inc patch, effectively removing the RC.
+		versionInfo = await runReleaseBranchCheckoutStep(
+			gitWorkingDirectory,
+			abortMessage,
+			semver.inc( packageVersion, 'patch' )
+		);
+	}
+
+	const { version, releaseBranch } = versionInfo;
```
Comment (ockham) [inline bin/plugin/commands/release.js:712]: See https://github.com/WordPress/gutenberg/pull/25971/commits/069f5f8cc1c0e6b3d981ad74744f8b9ee81b1fd1#r505636588 -- let's not obtain `version` from the return value of `runReleaseBranchCreationStep`/`runReleaseBranchCheckoutStep`, but simply set a `version` var inside of our `if`/`else` construct, and use that both as an arg _to_ those functions, _and_ here.

We might consider doing the same for `releaseBranch` -- we might just compute that value in this function now, and pass it to `runReleaseBranchCreationStep`/`runReleaseBranchCheckoutStep`.

File: bin/plugin/commands/release.js (hunk 10 above)
Comment (fullofcaffeine) [reply bin/plugin/commands/release.js:148]: You're right, I removed versionLabel for this reason and this slipped through the cracks.

Comment (ockham) [reply bin/plugin/commands/release.js]: Thanks for fixing!

**Review by ockham — APPROVED** [review]
Thanks, this is testing well for me now :+1:

**Comment by ockham** [comment]
@youknowriad @mcsf Could we get your final blessing please? :pray:

**Comment by youknowriad** [comment]
Ship it, thanks for your work here :)
